### PR TITLE
fix: implement extensions with template refactor

### LINKED ISSUE #775: Can not deploy a windowscluster with aks-engine 32.1
**Is this a request for help?**:
Yes.

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
Issue

**What version of aks-engine?**:
32.1

**Kubernetes version**:
1.14.0-beta.1

**What happened**:
When I try to deploy a cluster using Kubernetes Windows Walkthrough (https://github.com/Azure/aks-engine/blob/master/docs/topics/windows.md) it fails in the 'az group deployment create' step.

Azure Error: InvalidTemplate
Message: Deployment template validation failed: 'The template resource '3129k8s000winrm' at line '1' and column '139628' is invalid. 'ApiVersion' property is invalid. The supported versions are '2019-03-01,2018-11-01,2018-09-01,2018-08-01,2018-07-01,2018-06-01,2018-05-01,2018-02-01,2018-01-01,2017-12-01,2017-08-01,2017-06-01,2017-05-10,2017-05-01,2017-03-01,2016-09-01,2016-07-01,2016-06-01,2016-02-01,2015-11-01,2015-01-01,2014-04-01-preview,2014-04-01,2014-01-01,2013-03-01,2014-02-26,2014-04'. Please see https://aka.ms/arm-template/#resources for usage details.'.

I've done this many times before without any problems using a previous version of aks-engine.

**What you expected to happen**:
To get a working kubenetescluster in Azure

**How to reproduce it** (as minimally and precisely as possible):
Deploy with this kubernetes-windows-complete.json:
{
    "apiVersion":  "vlabs",
    "properties":  {
                       "orchestratorProfile":  {
                                                   "orchestratorType":  "Kubernetes",
                                                   "orchestratorRelease":  "1.14",
                                                   "orchestratorVersion": "1.14.0-beta.1"
                                               },
                       "masterProfile":  {
                                             "count":  1,
                                             "dnsPrefix":  "paletteonline14",
                                             "vmSize":  "Standard_D2_v3"
                                         },
                       "agentPoolProfiles":  [
                                                 {
                                                     "name":  "windowspool2",
                                                     "count":  2,
                                                     "vmSize":  "Standard_D2_v3",
                                                     "availabilityProfile":  "AvailabilitySet",
                                                     "osType":  "Windows",
                                                     "osDiskSizeGB":  128,
                                                     "extensions":  [
                                                                        {
                                                                            "name":  "winrm"
                                                                        }
                                                                    ]
                                                 }
                                             ],
                       "windowsProfile":  {
                                              "adminUsername":  "username",
                                              "adminPassword":  "password",
                                              "sshEnabled":  true
                                          },
                       "linuxProfile":  {
                                            "adminUsername":  "azureuser",
                                            "ssh":  {
                                                        "publicKeys":  [
                                                                           {
                                                                               "keyData":  *****************************
                                                                           }
                                                                       ]
                                                    }
                                        },
                       "servicePrincipalProfile":  {
                                                       "clientId":  ***********************************,
                                                       "secret":  **********************************
                                                   },
                       "extensionProfiles":  [
                                                 {
                                                     "name":  "winrm",
                                                     "version":  "v1"
                                                 }
                                             ]
                   }
}

### LINKED ISSUE #1079: Extensions are ignored for both master and agent nodes
**ISSUE**

---

**What version of aks-engine?**:

v0.33.0 onwards 

---

Consider this template:

```
{
    "location": "westeurope",
    "name": "prow-e05f8716-60f1-11e9-95d1-0e02a3139140",
    "APIVersion": "vlabs",
    "properties": {
        "orchestratorProfile": {
            "orchestratorType": "Kubernetes",
            "orchestratorRelease": "1.14"
        },
        "masterProfile": {
            "count": 1,
            "distro": "ubuntu",
            "dnsPrefix": "prow-custom-test3",
            "vmSize": "Standard_D2_v3"
        },
        "agentPoolProfiles": [
            {
                "name": "windowspool1",
                "count": 2,
                "distro": "ubuntu",
                "vmSize": "Standard_D2s_v3",
                "osType": "Windows",
                "availabilityProfile": "AvailabilitySet",
                "osDiskSizeGB": 128,
		"extensions": [
		     {
			"name": "winrm"
		     }
		]
            }
        ],
        "linuxProfile": {
            "adminUsername": "azureuser",
            "ssh": {
                "publicKeys": [
                    {
                        "keyData": ""
                    }
                ]
            }
        },
        "windowsProfile": {
            "adminUsername": "azureuser",
            "adminPassword": "Passw0rd1234",
            "WindowsImageSourceUrl": "",
            "WindowsPublisher": "",
            "WindowsOffer": "",
            "WindowsSku": "",
            "windowsDockerVersion": "18.09.5",
        },
        "extensionProfiles": [
            {
                "name": "winrm",
                "version": "v1"
            }
        ]
    }
}
```

The resulting ARM templates will not contain any Custom Script resources for the indicated extensions ( The template only includes a Windows agent extension, but it applies to master and Linux agents as well )

As far as I can tell, this happens after the introduction of this templating refactor: https://github.com/Azure/aks-engine/pull/324

Previously, the templates for the extension were generated by calling [1][2]. Post refactor, [2] is of course dead code and not being used in the new templating model, however I find no call to [1] or anything similar in the new model.

[1] https://github.com/Azure/aks-engine/blob/206328a0270c91b8b3dfa6d859eeff56b2f9c51c/pkg/engine/engine.go#L956
[2] https://github.com/Azure/aks-engine/blob/master/pkg/engine/template_generator.go#L625

Something similar to: https://github.com/adelina-t/aks-engine/commit/0b9bb97bb218d80d7e0d37d26b43df3eabd81de5 can be a quick fix for the issue, however not necessarily suited for permanence.

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**Reason for Change**:

Custom extensions were not implemented in the template refactor and left as a //todo. This PR links the existing extensions code in the template generator code.

**Issue Fixed**:

Fixes #1079 
Fixes #775 

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: pkg/engine/engine.go [hunk 1]
```diff
@@ -950,191 +946,6 @@ func getKubernetesPodStartIndex(properties *api.Properties) int {
 	return nodeCount + 1
 }
 
-// getLinkedTemplatesForExtensions returns the
-// Microsoft.Resources/deployments for each extension
-//func getLinkedTemplatesForExtensions(properties api.Properties) string {
-func getLinkedTemplatesForExtensions(properties *api.Properties) string {
```
Comment (CecileRobertMichon) [inline pkg/engine/engine.go]: copy pasted to vmextensions.go

File: pkg/engine/vmextensions.go [hunk 2]
```diff
@@ -206,3 +211,187 @@ func CreateAgentVMASAKSBillingExtension(cs *api.ContainerService, profile *api.A
 		VirtualMachineExtension: vmExtension,
 	}
 }
+
+// CreateLinkedTemplatesForExtensions returns the
+// Microsoft.Resources/deployments for each extension
+func CreateLinkedTemplatesForExtensions(properties *api.Properties) string {
```
Comment (CecileRobertMichon) [inline pkg/engine/vmextensions.go]: copy pasted from engine.go

File: pkg/engine/template_generator.go [hunk 3]
```diff
@@ -623,7 +623,7 @@ func (t *TemplateGenerator) getTemplateFuncMap(cs *api.ContainerService) templat
 			return DefaultInternalLbStaticIPOffset
 		},
 		"WriteLinkedTemplatesForExtensions": func() string {
-			extensions := getLinkedTemplatesForExtensions(cs.Properties)
+			extensions := CreateLinkedTemplatesForExtensions(cs.Properties)
```
Comment (CecileRobertMichon) [inline pkg/engine/template_generator.go]: we need to keep this for DCOS/Swarm

File: pkg/engine/vmextensions.go [hunk 4]
```diff
@@ -206,3 +206,47 @@ func CreateAgentVMASAKSBillingExtension(cs *api.ContainerService, profile *api.A
 		VirtualMachineExtension: vmExtension,
 	}
 }
+
+func CreateCustomExtensions(cs *api.ContainerService) []VirtualMachineExtensionARM {
```
Comment (CecileRobertMichon) [inline pkg/engine/vmextensions.go]: this is in a placeholder so far, in progress

File: pkg/engine/armvariables.go [hunk 5]
```diff
@@ -91,6 +91,7 @@ func getK8sMasterVars(cs *api.ContainerService) (map[string]interface{}, error)
 		"maximumLoadBalancerRuleCount":  maxLoadBalancerCount,
 		"masterFqdnPrefix":              "[tolower(parameters('masterEndpointDNSNamePrefix'))]",
 		"apiVersionCompute":             "2018-10-01",
+		"apiVersionDeployments":         "2018-05-01",
```
Comment (CecileRobertMichon) [inline pkg/engine/armvariables.go]: https://docs.microsoft.com/en-us/rest/api/resources/deployments

**Comment by CecileRobertMichon** [comment]
@adelina-t FYI, I think this works (was able to deploy with the apimodel in #1079 adding the rootURL to be my branch to fix the apiVersion issue). It also doesn't break Swarm or DCOS since it doesn't change the existing implementation.

File: pkg/engine/armvariables.go [hunk 6]
```diff
@@ -91,6 +91,7 @@ func getK8sMasterVars(cs *api.ContainerService) (map[string]interface{}, error)
 		"maximumLoadBalancerRuleCount":  maxLoadBalancerCount,
 		"masterFqdnPrefix":              "[tolower(parameters('masterEndpointDNSNamePrefix'))]",
 		"apiVersionCompute":             "2018-10-01",
+		"apiVersionDeployments":         "2018-06-01",
```
Comment (CecileRobertMichon) [inline pkg/engine/armvariables.go]: this essentially reverts #578 for extensions

File: pkg/engine/vmextensions.go [hunk 7]
```diff
@@ -206,3 +207,40 @@ func CreateAgentVMASAKSBillingExtension(cs *api.ContainerService, profile *api.A
 		VirtualMachineExtension: vmExtension,
 	}
 }
+
+func CreateCustomExtensions(properties *api.Properties) []DeploymentARM {
```
Comment (CecileRobertMichon) [inline pkg/engine/vmextensions.go]: todo: add unit test

File: pkg/engine/vmextensions.go [hunk 8]
```diff
@@ -206,3 +207,40 @@ func CreateAgentVMASAKSBillingExtension(cs *api.ContainerService, profile *api.A
...
+
+	for _, extensionProfile := range properties.ExtensionProfiles {
+		masterOptedForExtension, singleOrAll := validateProfileOptedForExtension(extensionProfile.Name, properties.MasterProfile.Extensions)
+		if masterOptedForExtension {
+			data, e := getMasterLinkedTemplateText(properties.MasterProfile, properties.OrchestratorProfile.OrchestratorType, extensionProfile, singleOrAll)
+			if e != nil {
+				fmt.Println(e.Error())
+			}
+			var ext DeploymentARM
+			if err := json.Unmarshal([]byte(data), &ext); err != nil {
```
Comment (CecileRobertMichon) [inline pkg/engine/vmextensions.go]: this doesn't work at the moment

File: pkg/engine/armvariables.go (hunk 6 above)
Comment (PatrickLang) [reply pkg/engine/armvariables.go]: how did you determine that version? I had a hard time finding out which docs to look at for the deployment versions.

**Comment by PatrickLang** [comment]
I'm hitting this:

```
+ az group deployment create --name plang0424 --resource-group plang0424rg --template-file ./_output/plang0424/azuredeploy.json --parameters ./_output/plang0424/azuredeploy.parameters.json
Azure Error: InvalidTemplate
Message: Deployment template validation failed: 'The template resource '[concat(variables('agentwinVMNamePrefix'), 'nic-', copyIndex(variables('agentwinOffset')))]' is not valid. Both api-version property at line '1' and column '71169' and api-profile property for the template cannot be null or empty. Either explicitly specify the api-version for the resource or specify the api-profile for the template. Please see https://aka.ms/arm-template/#resources for usage details.'.
```

I'll send you my apimodel privately

**Comment by PatrickLang** [comment]
Actually - here's an APImodel to test it.

```json
{
	"apiVersion": "vlabs",
	"properties": {
		"orchestratorProfile": {
			"orchestratorType": "Kubernetes"
		},
		"masterProfile": {
			"count": 1,
			"dnsPrefix": "",
			"vmSize": "Standard_D2_v2",
			"availabilityProfile": "AvailabilitySet"
		},
		"agentPoolProfiles": [
			{
				"name": "agentwin",
				"count": 2,
				"vmSize": "Standard_D2_v2",
				"availabilityProfile": "AvailabilitySet",
				"osType": "Windows",
				"extensions": [
					{
						"name": "windows-patches"
					}
				]
			}
		],
		"windowsProfile": {
			"adminUsername": "azureuser",
			"adminPassword": "replacepassword1234$",
			"enableAutomaticUpdates": false,
			"sshEnabled": true,
			"windowsOffer": "WindowsServerSemiAnnual",
			"windowsSku": "Datacenter-Core-1809-with-Containers-smalldisk",
			"imageVersion": "1809.0.20190410"
		},
		"linuxProfile": {
			"adminUsername": "azureuser",
			"ssh": {
				"publicKeys": [
					{
						"keyData": ""
					}
				]
			}
		},
		"extensionProfiles": [
			{
				"name": "windows-patches",
				"version": "v1",
				"rootURL": "https://raw.githubusercontent.com/Azure/aks-engine/master/",
				"extensionParameters": "'https://acs-mirror.azureedge.net/wink8s/Windows10.0-KB900259-x64-InstallForTestingPurposesOnly.exe'"
			}
		],
		"servicePrincipalProfile": {
			"clientId": "",
			"secret": ""
		}
	}
}
```

**Comment by PatrickLang** [comment]
and another template https://github.com/kubernetes-sigs/windows-testing/blob/master/job-templates/kubernetes_release.json

**Comment by CecileRobertMichon** [comment]
@PatrickLang this is still work in progress, it's not ready yet. I'm also getting the same error, working on it.

File: pkg/engine/armvariables.go (hunk 6 above)
Comment (CecileRobertMichon) [reply pkg/engine/armvariables.go]: the docs point to `2018-05-01` https://docs.microsoft.com/en-us/rest/api/resources/deployments but I was running into errors with that version so I used the one that was known working before https://github.com/Azure/aks-engine/pull/578/files#diff-aa4ff2b7c6ff5dfd7e2014283a7aa696L99.

File: pkg/engine/armtype.go [hunk 9]
```diff
@@ -26,6 +27,10 @@ type ARMResource struct {
 	DependsOn  []string          `json:"dependsOn,omitempty"`
 }
 
+// DeploymentARMResource is an alias for the ARMResource type to avoid MarshalJSON override
+type DeploymentARMResource ARMResource
```
Comment (CecileRobertMichon) [inline pkg/engine/armtype.go]: doing this fixes the stack overflow panic due to the  `MarshalJSON` override

File: pkg/engine/vmextensions_test.go [hunk 10]
```diff
@@ -367,3 +368,132 @@ func TestCreateAgentVMASCustomScriptExtension(t *testing.T) {
...
+	}
+	properties = &api.Properties{
+		OrchestratorProfile: &api.OrchestratorProfile{
+			OrchestratorType: Kubernetes,
+		},
+		ExtensionProfiles: []*api.ExtensionProfile{
+			{
+				Name:    "hello-world-k8s",
+				Version: "v1",
+				RootURL: "https://raw.githubusercontent.com/CecileRobertMichon/aks-engine/fix-extensions/",
```
Comment (CecileRobertMichon) [inline pkg/engine/vmextensions_test.go]: I'll have to replace this in a follow up PR in order to make this unit test pass both in this PR and in master once it merges (the template gets downloaded from github and master has a different apiVersion variable).

**Comment by CecileRobertMichon** [comment]
@PatrickLang this is ready now. I ran some tests and it worked fine, let me know if you run into issues.

**Comment by jackfrancis** [comment]
@CecileRobertMichon Should we add a "hello world" extension into the api model that we test on PRs?

**Comment by CecileRobertMichon** [comment]
@jackfrancis I don't think that's necessary: the unit tests test that we have the right ARM template so adding to E2E would be mostly testing ARM's deployment functionality since we don't test the functionality of the actual extension. I vote we prefer unit tests over E2E in this case.

File: docs/topics/extensions.md [hunk 11]
```diff
@@ -179,7 +179,7 @@ The following is an example of the template.json file.
    },
    "resources": [
     {
-      "apiVersion": "[parameters('apiVersionStorage')]",
+      "apiVersion": "[parameters('apiVersionDeployments')]",
```
Comment (jackfrancis) [inline docs/topics/extensions.md]: I think all the instances of `parameters('apiVersionDeployments')` need to be changed to `variables('apiVersionDeployments')`, right? I assume the `parameters('apiVersionStorage')` are docs cruft that wasn't cleaned up when some unit of params/vars rationalization work was completed.

Comment (CecileRobertMichon) [reply docs/topics/extensions.md]: good point, I was doing some automated search and replace. I'll change these to variables.

Comment (CecileRobertMichon) [reply docs/topics/extensions.md]: actually no sorry, they _are_ parameters. This is a deployment template with its own parameters defined.

Comment (CecileRobertMichon) [reply docs/topics/extensions.md]: https://github.com/Azure/aks-engine/pull/1133/files#diff-433cd94fef7d8fc09ee6e9ca8379b874R138

**Pull request merged**